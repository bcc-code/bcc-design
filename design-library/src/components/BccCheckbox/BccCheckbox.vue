<script setup lang="ts">
import { useId } from "@/hooks/use-id";
import { onMounted, ref, toRefs, watch } from "vue";

type Props = {
  value?: string | number | boolean;
  label?: string;
  disabled?: boolean;
  indeterminate?: boolean;
  size?: "base" | "lg" | "xl";
};

const props = withDefaults(defineProps<Props>(), {
  disabled: false,
  indeterminate: false,
  value: true,
  size: "base",
});

const modelValue = defineModel<boolean | Array<any>>();

const { indeterminate } = toRefs(props);

const input = ref<HTMLInputElement | null>(null);
const id = `bcc-checkbox-${useId()}`;

onMounted(() => {
  watch(
    indeterminate,
    (value) => {
      if (input.value) {
        const element = input.value;
        element.indeterminate = value;
      }
    },
    { immediate: true }
  );
});
</script>

<template>
  <div
    class="bcc-checkbox-wrapper"
    :class="{
      'bcc-checkbox-wrapper-disabled': disabled,
    }"
  >
    <input
      type="checkbox"
      class="bcc-checkbox"
      :class="size"
      :id="id"
      :disabled="disabled"
      v-model="modelValue"
      :value="value"
      ref="input"
    />
    <label :for="id" v-if="label">{{ label }}</label>
  </div>
</template>
