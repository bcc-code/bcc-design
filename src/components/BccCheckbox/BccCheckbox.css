@layer components {
    .bcc-checkbox {
        @apply form-checkbox rounded border-button-primary text-interactive focus:ring-2 focus:ring-emphasis;
        @apply disabled:bg-neutral-50 disabled:border-on-secondary disabled:hover:bg-neutral-50 disabled:hover:border-on-secondary;
    }
    .bcc-checkbox:checked:disabled, .bcc-checkbox:indeterminate:disabled {
        @apply bg-neutral-400 border-neutral-400;
    }

    /* Custom icons */
    .bcc-checkbox:checked {
        /* Generated with https://github.com/tigt/mini-svg-data-uri CLI */
        background-image: url("data:image/svg+xml,%3csvg viewBox='0 0 16 16' fill='white' xmlns='http://www.w3.org/2000/svg'%3e %3cpath d='M6.45458 9.84532L11.755 4.56004C11.8984 4.41869 12.068 4.34802 12.2639 4.34802C12.4597 4.34802 12.6283 4.41896 12.7697 4.56084C12.911 4.70271 12.9817 4.87334 12.9817 5.07271C12.9817 5.27208 12.911 5.44445 12.7697 5.58984L6.96948 11.39C6.82446 11.5314 6.65191 11.602 6.45182 11.602C6.25174 11.602 6.08102 11.5314 5.93968 11.39L3.21373 8.66408C3.07238 8.51802 3.00171 8.34522 3.00171 8.14566C3.00171 7.94609 3.07265 7.77563 3.21453 7.63429C3.3564 7.49294 3.52702 7.42227 3.7264 7.42227C3.92577 7.42227 4.09815 7.49294 4.24353 7.63429L6.45458 9.84532Z' fill='white'/%3e %3c/svg%3e")
    }
    .bcc-checkbox:indeterminate {
        /* Generated with https://github.com/tigt/mini-svg-data-uri CLI */
        background-image: url("data:image/svg+xml,%3csvg viewBox='0 0 16 16' fill='white' xmlns='http://www.w3.org/2000/svg'%3e %3cpath d='M4.12412 8.70615C3.9263 8.70615 3.75915 8.63766 3.62268 8.50068C3.48621 8.3637 3.41797 8.19592 3.41797 7.99736C3.41797 7.79878 3.48621 7.63188 3.62268 7.49666C3.75915 7.36143 3.9263 7.29382 4.12412 7.29382H11.8768C12.0746 7.29382 12.2443 7.36231 12.3857 7.49929C12.5272 7.63627 12.598 7.80405 12.598 8.00261C12.598 8.20119 12.5272 8.36809 12.3857 8.50331C12.2443 8.63854 12.0746 8.70615 11.8768 8.70615H4.12412Z' fill='white'/%3e %3c/svg%3e")
    }

    /* Wrapper */
    .bcc-checkbox-wrapper {
        @apply text-body-small flex items-center gap-2 text-secondary;
    }
    .bcc-checkbox-wrapper-disabled {
        @apply text-tertiary;
    }
}
