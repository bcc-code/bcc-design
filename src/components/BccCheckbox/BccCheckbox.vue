<script lang="ts">
export default {
  inheritAttrs: false,
};
</script>

<script setup lang="ts">
import { computed, onMounted, ref, toRefs, watch, type StyleValue } from "vue";
import { useAttrsWithoutStyles } from "@/composables/attrsWithoutStyles";
import { useId } from "@/hooks/use-id";

type Props = {
  modelValue: boolean;
  label?: string;
  disabled?: boolean;
  indeterminate?: boolean;
};

const props = withDefaults(defineProps<Props>(), {
  disabled: false,
  indeterminate: false,
});

const { modelValue, indeterminate } = toRefs(props);

const input = ref<HTMLInputElement | null>(null);
const id = `bcc-checkbox-${useId()}`;

const emit = defineEmits(["update:modelValue"]);

const checked = computed({
  get() {
    return !!modelValue.value;
  },
  set(value: boolean) {
    emit("update:modelValue", value);
  },
});

onMounted(() => {
  watch(
    indeterminate,
    (value) => {
      if (input.value) {
        const element = input.value;
        element.indeterminate = value;
      }
    },
    { immediate: true }
  );
});

const { attrs, attrsWithoutStyles } = useAttrsWithoutStyles();

const wrapperClasses = computed(() => {
  const classes = [];
  if (props.disabled) {
    classes.push("bcc-checkbox-wrapper-disabled");
  }
  if (attrs["class"]) {
    classes.push(attrs["class"]);
  }
  return classes;
});
</script>

<template>
  <div class="bcc-checkbox-wrapper" :class="wrapperClasses" :style="$attrs['style'] as StyleValue">
    <input
      type="checkbox"
      class="bcc-checkbox"
      :id="id"
      :disabled="disabled"
      v-model="checked"
      v-bind="attrsWithoutStyles"
      ref="input"
    />
    <label :for="id" v-if="label">{{ label }}</label>
  </div>
</template>
