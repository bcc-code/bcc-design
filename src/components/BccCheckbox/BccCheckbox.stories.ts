import BccCheckbox from "./BccCheckbox.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

/**
 * A checkbox input conforming to the design system styles
 */
export default {
  title: "Forms/BccCheckbox",
  component: BccCheckbox,
  argTypes: {},
} as Meta<typeof BccCheckbox>;

const Template: StoryFn<typeof BccCheckbox> = (args) => ({
  components: { BccCheckbox },
  setup() {
    return { args };
  },
  template: `
    <BccCheckbox v-bind="args" />
  `,
});

/**
 * Use `v-model` to set an initial boolean value for the checkbox.
 */
export const Example = Template.bind({});
Example.args = {
  label: "Example label",
  modelValue: true,
  disabled: false,
};
Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccCheckbox label="Example label" v-model="example" />      
`,
    },
  },
};

/**
 * Set the `disabled` prop to disable changing the checkbox. Set the `indeterminate` prop to show an indeterminate state (this doesn't affect the value).
 */
export const State: StoryFn<typeof BccCheckbox> = () => ({
  components: { BccCheckbox },
  template: `
    <div class="inline-flex flex-col space-y-4">
      <BccCheckbox label="Unchecked" :modelValue="false" />
      <BccCheckbox label="Checked" :modelValue="true" />
      <BccCheckbox label="Disabled unchecked" :modelValue="false" disabled />
      <BccCheckbox label="Disabled checked" :modelValue="true" disabled />
      <BccCheckbox label="Checked" :modelValue="true" />
      <BccCheckbox label="Indeterminate" :modelValue="true" indeterminate />
      <BccCheckbox label="Indeterminate disabled" :modelValue="true" indeterminate disabled />
    </div>
  `,
});
