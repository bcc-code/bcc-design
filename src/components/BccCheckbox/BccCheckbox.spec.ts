import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import BccCheckbox from "./BccCheckbox.vue";

describe("BccCheckbox", () => {
  it("renders a checked or unchecked checkbox", () => {
    const wrapper = mount(BccCheckbox, { props: { modelValue: false } });
    expect(wrapper.html()).toMatchSnapshot();

    wrapper.setProps({ modelValue: true });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders a label", () => {
    const wrapper = mount(BccCheckbox, {
      props: { label: "Test label", modelValue: false },
    });

    expect(wrapper.text()).toBe("Test label");
    expect(wrapper.html()).toMatchSnapshot();
  });
});
