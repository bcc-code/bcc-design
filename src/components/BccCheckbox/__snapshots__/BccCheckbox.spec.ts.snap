// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccCheckbox > renders a checked or unchecked checkbox 1`] = `"<div class=\\"bcc-checkbox-wrapper\\"><input type=\\"checkbox\\" class=\\"bcc-checkbox\\" id=\\"bcc-checkbox-1\\"><label for=\\"bcc-checkbox-1\\"></label></div>"`;

exports[`BccCheckbox > renders a checked or unchecked checkbox 2`] = `"<div class=\\"bcc-checkbox-wrapper\\"><input type=\\"checkbox\\" class=\\"bcc-checkbox\\" id=\\"bcc-checkbox-1\\"><label for=\\"bcc-checkbox-1\\"></label></div>"`;

exports[`BccCheckbox > renders a label 1`] = `"<div class=\\"bcc-checkbox-wrapper\\"><input type=\\"checkbox\\" class=\\"bcc-checkbox\\" id=\\"bcc-checkbox-2\\"><label for=\\"bcc-checkbox-2\\">Test label</label></div>"`;
